### PR TITLE
OSRA380 - View and Controller for show action - Orphans

### PR DESCRIPTION
**Pull request by gmunumel**
OSRA380 - View and Controller for show action - Orphans

### CODE REVIEW

File: app/views/hq/orphans/show.html.haml [hunk 1]
```diff
@@ -0,0 +1,127 @@
+%div.action_items
+  %div.instance
+    = link_to 'Edit Orphan', edit_hq_orphan_path(@orphan.id),
+      class: 'btn btn-default', role: 'button'
+    = link_to 'Link to Orphan', '#',
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: In the ActiveAdmin interface, creation of sponsorships takes place only from a sponsor's show view. Let's keep it the same in HQ for now, and make the change if the client requests it.

File: app/views/hq/orphans/show.html.haml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    = @orphan.minor_siblings_count
+%dl.dl-horizontal
+  %dt Sponsored Minor Siblings Count
+  %dd
+    = @orphan.sponsored_minor_siblings_count
+%dl.dl-horizontal
+  %dt Original Province Name
+  %dd
+    = @orphan.original_address.province.name
+%dl.dl-horizontal
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: It seems that there is no need to display province codes here

File: app/views/hq/orphans/show.html.haml [hunk 3]
```diff
@@ -0,0 +1,127 @@
+%div.action_items
+  %div.instance
+    = link_to 'Edit Orphan', edit_hq_orphan_path(@orphan.id),
+      class: 'btn btn-default', role: 'button'
+    = link_to 'Link to Orphan', '#',
+      class: 'btn btn-default', role: 'button'
+
+%dl.dl-horizontal
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: Let's break this long display of attributes into thematic panels, the way it's currently done in the AA interface. Have a look at http://getbootstrap.com/components/#panels

File: app/views/hq/orphans/show.html.haml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  %dt Minor Siblings Count
+  %dd
+    = @orphan.minor_siblings_count
+%dl.dl-horizontal
+  %dt Sponsored Minor Siblings Count
+  %dd
+    = @orphan.sponsored_minor_siblings_count
+%dl.dl-horizontal
+  %dt Original Province Name
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: Once these fields are wrapped into 'Original Address' and 'Current Address' panels, the words 'Original' and 'Current' can be omitted from individual labels.

File: app/views/hq/orphans/show.html.haml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    = @orphan.original_address.neighborhood
+%dl.dl-horizontal
+  %dt Original Street
+  %dd
+    = @orphan.original_address.street
+%dl.dl-horizontal
+  %dt Current Province Name
+  %dd
+    = @orphan.current_address.province.name
+%dl.dl-horizontal
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: As above

File: app/views/hq/orphans/show.html.haml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    = @orphan.current_address.neighborhood
+%dl.dl-horizontal
+  %dt Current Street
+  %dd
+    = @orphan.current_address.street
+%dl.dl-horizontal
+  %dt Name Status
+  %dd
+    = @orphan.orphan_status.name
+%dl.dl-horizontal
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: Don't need the code

File: app/views/hq/orphans/show.html.haml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    = @orphan.orphan_status.code
+%dl.dl-horizontal
+  %dt Priority
+  %dd
+    = @orphan.priority
+%dl.dl-horizontal
+  %dt Sponsorship Status
+  %dd
+    = @orphan.orphan_sponsorship_status.name
+%dl.dl-horizontal
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: As above

File: app/views/hq/orphans/show.html.haml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  %dt Current Neighborhood
+  %dd
+    = @orphan.current_address.neighborhood
+%dl.dl-horizontal
+  %dt Current Street
+  %dd
+    = @orphan.current_address.street
+%dl.dl-horizontal
+  %dt Name Status
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: Just 'Status' or 'Orphan Status'. I would be more natural to display these attributes next to others that pertain directly to the orphan. See the current AA interface - the grouping of fields into panels there feels logical and is client-approved.

File: spec/controllers/hq/orphans_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+require 'will_paginate/array'
+
+RSpec.describe Hq::OrphansController, type: :controller do
+    let(:orphan) {build_stubbed :orphan}
+    let(:orphans) {(1..5).map {build_stubbed :orphan}}
```
Comment (NikitaAvvakumov) [inline spec/controllers/hq/orphans_controller_spec.rb]: Don't need this here until `#index` is implemented

File: spec/views/hq/orphans/show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+RSpec.describe "hq/orphans/show.html.haml", type: :view do
+  let(:orphans) { FactoryGirl.build_stubbed(:orphan_list) }
```
Comment (NikitaAvvakumov) [inline spec/views/hq/orphans/show_spec.rb]: The orphan factory builds the `orphan_list` association. Do you need to create it here explicitly?

File: spec/views/hq/orphans/show_spec.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  let(:orphan) { FactoryGirl.build_stubbed(:orphan, orphan_list: orphans) }
+
+  describe 'the orphan exists' do
+    before :each do
+      assign(:orphan, orphan)
```
Comment (NikitaAvvakumov) [inline spec/views/hq/orphans/show_spec.rb]: This is a minor style point: when you need to use a variable in the spec and also have it assigned to an instance variable in the view being tested, you can replace the lines
```ruby
let(:orphan) { build_stubbed :orphan }
before do
  assigns :orphan, orphan
  render
end
```
with
```ruby
before do
  @orphan = build_stubbed :orphan
  render
end
```
The `@orphan` variable is then used in both the view and the spec (you would need to change all references to `orphan` to `@orphan` in the `it` blocks).

**Comment by gmunumel** [comment]
@NikitaAvvakumov Thanks for the clarification. I will try to made all these changes

File: spec/views/hq/orphans/show_spec.rb (hunk 11 above)
Comment (gmunumel) [inline spec/views/hq/orphans/show_spec.rb]: Yes, you are right. I try to follow the same idea of `sponsor`. But for orphan is more simple.

File: app/views/hq/orphans/show.html.haml (hunk 8 above)
Comment (gmunumel) [inline app/views/hq/orphans/show.html.haml]: I can group and try to mimic the look of AA interface, for instance, made a panel for orphan details, parent's details, etc. Is it ok?

Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: We don't need to follow the AA look verbatim, but with such a long list of attributes some kind of grouping into categories is needed. You can take AA as a starting point, but feel free to deviate from it if something doesn't feel right to you.

**Comment by gmunumel** [comment]
Screenshot of the orphan right now.

screenshot from 2015-02-24 20 10 29

**Comment by gmunumel** [comment]
Do I need to write behaviors test for this?

**Comment by NikitaAvvakumov** [comment]
@gmunumel We haven't written integration tests for code like this so far, though I'm starting to wonder whether a very high-level feature can be created to test the entirety of the CRUD operations for a particular model - e.g.
```cucumber
Scenario: admin user manipulates orphan data
  Given an orphan record exists
  Then I should be able to look at it
  And I should be able to change it
  And I should be able to see it in a list of all orphan records
```
What do people think?

**Comment by gmunumel** [comment]
I think sooner or later we will need them. :+1: But we can plan to add them in the next sprint.

File: app/views/hq/orphans/show.html.haml [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      %dt Sponsorship Status
+      %dd
+        = @orphan.orphan_sponsorship_status.name
+    %dl.dl-horizontal
+      %dt Priority
+      %dd
+        = @orphan.priority
+.panel.panel-default
+  .panel-heading 
+    %h3.panel-title Parent's Details
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: "Parents' details"

File: app/views/hq/orphans/show.html.haml [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  .panel-body
+    %dl.dl-horizontal
+      %dt Father Given Name
```
Comment (NikitaAvvakumov) [inline app/views/hq/orphans/show.html.haml]: Let's take this opportunity to correct the app's grammar: "Father's Given Name" & "Mother's Given Name". This has bugged me for a long time :)

**Comment by razorcd** [comment]
@NikitaAvvakumov : yes, for the CRUD operations we ca do that

File: app/views/hq/orphans/show.html.haml (hunk 13 above)
Comment (gmunumel) [inline app/views/hq/orphans/show.html.haml]: sure:+1:

**Pull request merged**